train_gpt2.cu

```diff
@@ -997,12 +1090,21 @@ float gpt2_calculate_grad_norm(GPT2 *model, MultiGpuConfig* multi_gpu_config) {
         // further sum the (partial) squared norm across all GPUs
         ncclCheck(ncclAllReduce(grad_norm_squared, grad_norm_squared, sizeof(float), ncclFloat, ncclSum, multi_gpu_config->nccl_comm, main_stream));
 #endif
-    } else {
```

Comment:
I generally would try to keep `#if` guards to a minimum, they just make it harder to reason about the code because you end up compiling different source code. So I'm trying to only hide code behind `#if` if it actually cannot compile in single-gpu mode, e.g., because we don't have the nccl_stream. Ideally, I think all that should be hidden inside `zero.h`.